LoopKitUI/UIResult.swift

```diff
@@ -0,0 +1,13 @@
+//
+//  UIResult.swift
+//  LoopKitUI
+//
+//  Created by Darin Krauss on 1/21/21.
+//  Copyright © 2021 LoopKit Authors. All rights reserved.
+//
+
+public enum UIResult<UserInteractionRequired, Success, Failure> where Failure: Error {
+    case userInteractionRequired(UserInteractionRequired)
+    case success(Success)
+    case failure(Failure)
```

Comment:
I'll agree this isn't ideal. To be honest, it was an attempt to improve upon the following type of pattern from the previous implementation (in this case from `CGMManagerUI.swift`):

```
    /// If this method returns nil, it's expected that `init?(rawState: [:])` creates a non-nil manager
    static func setupViewController(glucoseTintColor: Color, guidanceColors: GuidanceColors) -> (UIViewController & CGMManagerSetupViewController & CompletionNotifying)?
```

The way I read this, on one-hand, we are giving the CGMManager implementation the ability to return some custom UI which ultimately creates an instance of a CGMManager any way it wants. Loop doesn't know and doesn't care. I just gets a notification that a CGMManager was created (via `CGMManagerSetupViewController` here). I'd say this is good and gives all the flexibility to the CGMManager implementation.

On the other hand, we are saying if no custom UI is needed then the CGMManager implementation has zero-choice about how it is instantiated and it *must* support `init(rawValue:)` with an empty dictionary. Requiring support for `init(rawValue:)` is okay since it already must do so for rehydration, but using the empty dictionary as a flag to say "create one from scratch" and that it *must* return non-nil sounds a lot like it should just support `init()`.

In either case (`init()` or `init(rawValue:)` with an empty dictionary), though, it removes all flexibility/control from the CGMManager implementation about how it is instantiated if a UI is not needed.

Also, earlier in this PR you commented against the non-descriptive nature of `init()` in the `Onboarding` protocol above. I think this would follow similarly for the `init(rawValue:)` with an empty dictionary route.

If we want to stick with the name `setupViewController` that returns an optional UI, then it would follow we have a counterpart `setupWithoutUI` (or similar) static func that is used to instantiate the CGMManager in case UI is not required and it would be on the `CGMManager` protocol (not `CGMManagerUI`). I'd suggest the signature look something like:

```
    static func setupWithoutUI() -> Result<CGMManager, Error>
```

Returning an error allows the implementation to indicate that no, a UI is required and the counterpart `setupViewController` on `CGMManagerUI` should be invoked.

Thoughts?

Reply:
I'm fine returning an enum that can be either a UI, or a created CGMManager.  It doesn't seem like we need an error state here, so I'm opposed to that.

Reply:
Okay, I will do that. I will create an different enum for each of CGM manager, pump manager, and service that returns either a conforming view controller or instantiated object (CGM manager, pump manager, or service).

Reply:
I think a templatized enum shared between them as you have now is fine, my only issues are with the error state (remove it) and the name (make it more specific: `SetupUIResult` or something)

Reply:
Got it. Will do. Thanks.